Implementation of msgcat and msgmerge utilities from GNU gettext

soft-suroleb opened the pull request:
### Implementation of some features for msgcat and msgmerge to work with a compendium in PyBabel

#### pybabel concat

- Allows concatenating multiple .po files into one.
- In case of translation conflicts, the translation from the first file is taken, as if the use-first option is set to true. Flags, locations, and other information in messages are combined.
- Implemented options:
  - output-file
  - less-than
  - more-than
  - unique
  - use-first
  - no-location
  - width
  - no-wrap
  - sort-output
  - sort-by-file

#### pybabel merge

- Enables merging files using a compendium for translation memory.
- The compendium can be used in two modes:
  - Default mode: Translations from the compendium are taken if absent in the output file.
  - Compendium overwrite mode: With the compendium-overwrite option, translations in the compendium are considered primary and overwrite those in the output file. If a translation is taken from the compendium, a comment is added specifying the source.
- Implemented options:
  - input-files
  - compendium
  - compendium-overwrite
  - no-compendium-comment
  - update
  - output-file
  - backup
  - suffix
  - no-fuzzy-matching
  - no-location
  - width
  - no-wrap
  - sort-output
  - sort-by-file

soft-suroleb committed:
Add definitions for MessageConcatenation and MessageMerge classes in pybabel

 * Define the MessageConcatenation class to mimic the functionality of GNU gettext's msgcat

 * Define the MessageMerge class to mimic the functionality of GNU gettext's msgmerge

 * Implement placeholders for the main interface functions

soft-suroleb committed:
Implement basic logic for concatenating catalogs

 * Add validation for main msgcat options - input_files, output_file

 * Temporarily set use_first option to true to avoid handling cases with different translations for the same messages

soft-suroleb committed:
Add options: unique, less-than, more-than, no-wrap, and width

 * Implement options unique, less-than, and more-than, and validate their dependencies with each other.
   * These options specify which messages to include in the output file.

 * Implement and validate options no-wrap and width.

 * Create a helper function _prepare that collects data on message occurrences across different catalogs.

 * Mark options that are already implemented #

soft-suroleb committed:
Implement basic msgmerge logic for working with a compendium

 * Implement basic functionality of msgmerge

 * Use and validate the main options: input-files and output-file

 * Use and validate options: no-wrap and width

 * Use and validate options: sort-output and sort-by-file, both in msgmerge and msgcat

 * In the basic version of working with a compendium, a translation for a message is taken from the compendium only if the resulting catalog lacks a translation.

soft-suroleb committed:
Write tests for msgcat

 * Create basic tests to verify the functionality of msgcat, specifically
   the concatenation of catalogs, merging of message flags, locations, etc.

 * Remove the validation of options sort-output, sort-by-file, unique, use-first,
   as they are initialized in the function initialize_options.

soft-suroleb committed:
Write tests for msgmerge

 * Create basic tests to verify the functionality of msgmerge, specifically the merging of messages and their integration with a compendium.

 * Remove the definition of sort-output and sort-by-file, and add an additional check for input-files.

soft-suroleb committed:
Add options update, backup, and c_overwrite for a different compendium handling logic

 * Implement `update` to update the source file instead of writing to the current output file

 * Implement `backup` to save a backup of the source file before making any updates

 * Implement `c_overwrite` to use a new mode of handling the compendium, where translations from the compendium overwrite messages in the output file

soft-suroleb committed:
Add test for msgmerge compendium overwrite mode with no comments

 * Implement a test for `msgmerge` that validates the new mode where compendium entries overwrite messages in the output PO file.

 * Include the `no_compendium_comment` option to ensure comments about translations sourced from the compendium are not included.

 * Utilize the `no-location` option to exclude location comments from the output.

soft-suroleb committed:
Refactor test for msgmerge with compendium-overwrite option

 * Implemented a helper function `_get_expected` to standardize the expected PO file structure.
 * Renamed the option `c-overwrite` to `compendium-overwrite`

soft-suroleb committed:
Create a catalog without fuzzy by default, remove add-location

 * Mark the catalog as fuzzy after msgcat and msgmerge if there is at least one fuzzy message

 * Remove add-location as it's unnecessary

soft-suroleb committed:
Rename msgmerge to merge and msgcat to concat

soft-suroleb committed:
Add discription to all options

soft-suroleb commented:
Hi guys! I see you marked pr but still haven't left any comments. Will it be reviewed?

soft-suroleb committed:
Ability to specify multiple compendiums

soft-suroleb committed:
Marking conflicting messages

* Update _prepare function in ConcatenateCatalog to check conflicting messages and to not parse po-files twice

* Add _conflicts field in Catalog to mark conflicts

* Update tests

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,318 @@ def run(self):
+        self.properties_input = None
+        self.stringtable_input = None
+        self.to_code = None
+        # the first translation is always used temporarily
```
What does this comment mean?

soft-suroleb replied:
When one message has different translations in different po-files, gnu gettext adds a note that there is a conflict/ I didn't implement this logic at first, but then i made a commit with conflict support, so i removed the comment

akx reviewed (changes requested):
Thank you! Some initial comments within.

soft-suroleb committed:
Fix PR issues

* Delete unused options

* Fix multiline options comments

* Replace backup logic in MergeCatalog

* Rename to ConcatenateCatalog

soft-suroleb commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,318 @@ def run(self):
+        ('input-files', None, 'input files'),
+        ('files-from=', 'f', 'get list of input files from FILE'),
+        ('directory=', 'D', 'add DIRECTORY to list for input files search'
+                            'If input file is -, standard input is read.'),
```
Removed as unused option

soft-suroleb commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,318 @@ def run(self):
+    def finalize_options(self):
+        if not self.input_files:
+            raise OptionError('you must specify the input files')
+        if not self.output_file:
+            raise OptionError('you must specify the output file')
```
Do you want the result to be output to stdout or what? It's just that each command in frontend has a similar exception for the output_file

soft-suroleb commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,318 @@ def run(self):
+            self.more_than = int(self.more_than)
+        if self.less_than is not None:
+            self.less_than = int(self.less_than)
+        if self.unique:
+            self.less_than = 2
```
Not quite so. More_than and less_than are like bounds. Unique means less_than=2 but boolean

soft-suroleb commented on tests/messages/test_frontend.py:
```diff
@@ -715,6 +715,431 @@ def test_supports_width(self):
         assert expected_content == actual_content
 
 
+class ConcatanationCatalogTestCase(unittest.TestCase):
```
Is a new file really needed here? The file is test_frontend.py and all other tests for commands are there and it's logical that they will be there too, no?

soft-suroleb commented:
Made support for conflicting messages from different po-files during concatenation. If one message has different translations in different files, then the corresponding strings marked with a comment about conflict

soft-suroleb committed:
Add info about pybable concat and pybabel merge into docs

soft-suroleb committed:
Add usage documentation for pybabel concat and merge commands

 * Includes .rst file with detailed use cases and practical examples for pybabel's concat and merge utilities, outlining common scenarios, options, and best practices for managing PO files.

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,247 @@ def run(self):
+        for filename in self.input_files:
+            with open(filename, 'r') as pofile:
+                template = read_po(pofile)
+            for message in template:
```
This should probably also filter out messages with a nullish `id`.

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,318 @@ def run(self):
+
+    def finalize_options(self):
+        if not self.input_files or len(self.input_files) != 2:
+            raise OptionError('must be two po files')
```
That would be useful information here in the error message too.

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,247 @@ def run(self):
+class MergeCatalog(CommandMixin):
+    description='updates translation PO file by merging them with updated template POT file with using compendium'
+    user_options=[
+        ('input-files', None, 'def.po (obsolete translations) ref.pot (actual template)'),
```
```suggestion
        ('input-files', None, 'exactly two input files: def.po (obsolete translations); ref.pot (current template)'),
```

I assume "actual" means "current" here (common [false friend](https://en.wikipedia.org/wiki/False_friend)) :)

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,247 @@ def run(self):
+        elif self.width is not None:
+            self.width = int(self.width)
+
+    def _get_message_from_compendium(self, compendium):
+        for file_path in compendium:
```
Even if the CLI argument is in the singular, it would be nicer to read here in the plural.

Also, this clearly returns things in the plural:

```suggestion
    def _get_messages_from_compendiums(self, compendium_paths):
        if not compendium_paths:
            return
        for file_path in compendium_paths:
```

akx commented on babel/messages/frontend.py:
```diff
@@ -852,6 +853,247 @@ def run(self):
+            no_fuzzy_matching=self.no_fuzzy_matching
+        )
+
+        if self.compendium:
+            for message, compendium_path in self._get_message_from_compendium(self.compendium):
+                current = catalog[message.id]
+                if message.id in catalog and (not current.string or current.fuzzy or self.compendium_overwrite):
+                    if self.compendium_overwrite and not current.fuzzy and current.string:
+                        catalog.obsolete[message.id] = current.clone()
+
+                    current.string = message.string
+                    if current.fuzzy:
+                        current.flags.remove('fuzzy')
+
+                    if not self.no_compendium_comment:
+                        current.auto_comments.append(compendium_path)
```
This could be unnested, and `get_messages_from_compendiums` could early-return if the passed argument is falsy.

```suggestion
        for message, compendium_path in self._get_message_from_compendium(self.compendium):
            current = catalog[message.id]
            if message.id in catalog and (not current.string or current.fuzzy or self.compendium_overwrite):
                if self.compendium_overwrite and not current.fuzzy and current.string:
                    catalog.obsolete[message.id] = current.clone()

                current.string = message.string
                if current.fuzzy:
                    current.flags.remove('fuzzy')

                if not self.no_compendium_comment:
                    current.auto_comments.append(compendium_path)
```

akx commented on docs/cmdline.rst:
```diff
@@ -247,3 +247,92 @@ filename of the output file will be::
+concat
+======
+
+The `concat` command merges multiple PO files into a single one. If a message has
```
```suggestion
The `concat` command merges multiple PO files into one. If a message has
```

akx commented on docs/cmdline.rst:
```diff
@@ -247,3 +247,92 @@ filename of the output file will be::
+different translations in different PO files, the conflicting translations are
+marked with a conflict comment::
+    #-#-#-#-#  <file> (PROJECT VERSION)  #-#-#-#-#
+and the message itself is marked with a `fuzzy` flag::
```
Is the `::` on purpose here?

akx commented on docs/cmdline.rst:
```diff
@@ -247,3 +247,92 @@ filename of the output file will be::
+
+The `merge` command allows updating files using a compendium as a translation memory::
+
+    $ pybabel concat --help
```
```suggestion
    $ pybabel merge --help
```

akx commented on docs/concat_merge_usage.rst:
```diff
@@ -0,0 +1,52 @@
+Usage scenarios
```
I think these could be in the `cmdline.rst` file – they'd probably be easier to find there.

akx commented on docs/cmdline.rst:
```diff
@@ -247,3 +247,92 @@ filename of the output file will be::
+  from the compendium take priority and replace those in the output file. If a translation
+  is used from the compendium, a comment noting the source is added
+
+The ``input-files`` option includes def.po, a file with obsolete translations, and ref.pot,
```
```suggestion
The ``input-files`` option accepts exactly two arguments: a file with obsolete translations, and
```

akx commented on docs/cmdline.rst:
```diff
@@ -247,3 +247,92 @@ filename of the output file will be::
+
+The ``compendium`` option can be specified multiple times to use several compendiums.
+
+The ``backup`` option is used to create a backup copy of the def.po file, which contains
+obsolete translations
+
+The ``suffix`` option allows you to specify a custom suffix for the backup file
+By default, a standard suffix ``~`` is appended to the backup file's name,
```
```suggestion
The ``backup`` option is used to create a backup copy of the def.po file, which contains
obsolete translations. The ``suffix`` option allows you to specify a custom suffix for the backup file (defaulting to ``~``).
```

akx commented on babel/messages/frontend.py:
```diff
@@ -866,13 +1108,17 @@ class CommandLineInterface:
         'init': 'create new message catalogs from a POT file',
         'update': 'update existing message catalogs from a POT file',
+        'concat': 'concatenates and merges the specified PO files',
+        'merge': 'combines two Uniforum-style PO files into one',
```
I still have no idea what an Uniforum-style PO file is; do we not support whatever is a non-Uniforum-style PO file?

akx commented on babel/messages/pofile.py:
```diff
@@ -642,6 +645,37 @@ def _format_comment(comment, prefix=''):
             yield f"#{prefix} {line.strip()}\n"
 
+    def _format_conflict_comment(file, project, version, prefix=''):
+        comment = f"#-#-#-#-#  {file} ({project} {version})  #-#-#-#-#"
```
Is this standard notation, or are we introducing a Babel-specific comment extension? If the latter, I'd rather not...

soft-suroleb replied:
This is the notation used in gnu gettext
/

akx reviewed (changes requested):
Another round of comments :)

soft-suroleb committed:
Fix PR issues

* Rename file_name to filename

* Adding fuzzy flag to message parameterized in 'add_conflict'

* Replace usage scenarious to cmdline.rst

* Rename to ConcatenateCatalog

soft-suroleb committed:
Add '_conflicts' getter in catalog

soft-suroleb commented:
Guys? Will there be another round? :)

akx commented:
@soft-suroleb Sorry, I was honeymooning in Japan for the last 3 weeks 😄 

There's still a bunch of unresolved comments from me from the last round of review, I think?

soft-suroleb commented:
Oh, my congrats :) Yeah, but there is question for you in the unresolved comment

soft-suroleb commented:
Guys?

FrBrGeorge commented:
I have to ask — how is it going? I'm waiting for this feature since it was implemented in 2024… Can I help to move it further somehow?

soft-suroleb committed:
Remove fuzzy flag mutation from add_conflict

The add_conflict method was mutating the message's fuzzy flag as a side
effect, which is unexpected. The caller (ConcatenateCatalog.run) now
explicitly sets the fuzzy flag when adding a conflicting message.

soft-suroleb committed:
Refactor ConcatenateCatalog: stdout output, mutual exclusivity, better help texts

soft-suroleb committed:
Fix MergeCatalog options and update docs

soft-suroleb committed:
Add tests for ConcatenateCatalog and MergeCatalog

soft-suroleb commented:
@akx @FrBrGeorge 
All review comments have been addressed:

- Removed fuzzy flag mutation from `add_conflict` — the caller now sets it explicitly
- Refactored `ConcatenateCatalog`: fixed help texts, output to stdout when no `--output-file` is given, added mutual exclusivity of `--unique` with `--less-than`/`--more-than`, replaced nested dict with `Counter` + `defaultdict`, renamed `_prepare` → `_collect_message_info`
- Fixed `MergeCatalog`: corrected description, fixed typo in `--update` help, added missing help texts, fixed `multiple_value_options` tuple (missing trailing comma), improved error message, renamed and refactored `_get_messages_from_compendiums` with early return and walrus operator
- Fixed `cmdline.rst`: corrected heading levels, typos, `pybabel concat` → `pybabel merge` in the merge section, updated help text examples

Also expanded the test suite with additional cases: stdout output, `--unique` mutual exclusivity, `--use-first` behaviour, and plural message handling.

akx commented on babel/messages/catalog.py:
```diff
@@ -780,6 +790,18 @@ def __setitem__(self, id: _MessageID, message: Message) -> None:
             self._messages[key] = message
 
+    def add_conflict(self, message: Message, filename: str, project: str, version: str):
+        key = message.id
```
Should this use `self._key_for` like for keying `self._messages` above? IOW, should keying take message context into account?

akx commented on babel/messages/frontend.py:
```diff
@@ -887,6 +888,262 @@ def run(self):
+
+    def _collect_message_info(self):
+        templates: list[tuple[str, Catalog]] = []
+        message_counts: Counter = Counter()
```
```suggestion
        message_counts = Counter()
```
The type hint doesn't add extra information :)

akx commented on babel/messages/frontend.py:
```diff
@@ -887,6 +888,262 @@ def run(self):
+    def _collect_message_info(self):
+        templates: list[tuple[str, Catalog]] = []
+        message_counts: Counter = Counter()
+        message_strings: dict[object, set] = defaultdict(set)
```
Could we have a tighter key type than `object`? Could the type of the set also be determined? (I'm looking at line 974, where it's hard for me to figure out what the actual set value type is.)

akx commented on babel/messages/frontend.py:
```diff
@@ -887,6 +888,262 @@ def run(self):
+
+    def _get_messages_from_compendiums(self, compendium_paths):
+        if not compendium_paths:
+            return
```
Unnecessary, unless you imply for `compendium_paths` to be None sometimes, which sounds unlikely?
```suggestion
```

akx commented on babel/messages/pofile.py:
```diff
@@ -351,8 +351,11 @@ def parse(self, fileobj: IO[AnyStr] | Iterable[AnyStr]) -> None:
+                if line[1:].startswith('-'):
+                    self._invalid_pofile(line, lineno, 'cannot parse po file with conflicts')
+
+                if line[1:].startswith('~'):
```
Please see https://github.com/python-babel/babel/commit/d7a7589a6cee3aa4c68de60f4d69a9cdad50a7ff - we're avoiding `.startswith` on purpose.

akx commented on babel/messages/pofile.py:
```diff
@@ -351,8 +351,11 @@ def parse(self, fileobj: IO[AnyStr] | Iterable[AnyStr]) -> None:
-                if line[:2] == '#~':
+            if line.startswith('#'):
+                if line[1:].startswith('-'):
+                    self._invalid_pofile(line, lineno, 'cannot parse po file with conflicts')
```
Should this be a warning rather than a hard error? Rather, how has Babel parsed files with these conflict markers in the past?

soft-suroleb replied:
_invalid_pofile is already a warning by default — it only raises when abort_invalid=True is explicitly passed to read_po().

Previously, #-#-#-#-# lines had no special handling: they silently became user comments, and the conflicted msgstr was parsed as regular text, producing a malformed catalog with no diagnostic. The current behavior is strictly better.

akx commented on babel/messages/catalog.py:
```diff
@@ -780,6 +790,18 @@ def __setitem__(self, id: _MessageID, message: Message) -> None:
+        })
+
+    def get_conflicts(self, id: _MessageID) -> list[ConflictInfo]:
+        return self._conflicts.get(id, [])
```
These new public APIs should be designed with care (e.g. should these actually take message instances?) and have docstrings.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+    with freeze_time("1994-11-11"):
+        yield
+
+
```
I'd appreciate it if this test file was written in Pytest style (see #1240, #1241 for the conversion that occurred recently) from the get-go?

Could I also ask for it to be split into `test_concat` and `test_merge`, like `test_compile` and `test_extract` already are?

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        yield
+
+
+class TestConcatanateCatalog:
```
Typo in suite name.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            catalog.add('almost_same', string='A bit same',  locations=[('hard.py', 1000)], flags=['flag3'])
+            catalog.add('same', string='Same', locations=[('hard.py', 100)], flags=['flag4'])
+            catalog.add(('plural', 'plurals'), string=('Plural', 'Plurals other'), locations=[('hard.py', 2000)])
+            pofile.write_po(file, catalog)
```
It looks like this wouldn't need to mess with `data_dir` at all, but just work in a Pytest `tmp_path`?

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+    def teardown_method(self):
+        for file in [self.temp1, self.temp2, self.output_file]:
+            if os.path.isfile(file):
+                    os.unlink(file)
```
This would become unnecessary using `tmp_path`.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        self.dist = Distribution(TEST_PROJECT_DISTRIBUTION_DATA)
+        self.cmd = frontend.ConcatenateCatalog(self.dist)
+        self.cmd.initialize_options()
```
`concat_cmd` could be a Pytest fixture.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        self.temp1 = f'{i18n_dir}/msgcat_temp1.po'
+        self.temp2 = f'{i18n_dir}/msgcat_temp2.po'
+        self.output_file = f'{i18n_dir}/msgcat.po'
```
These could be `pathlib.Path`s.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        with open(self.output_file, 'r') as f:
+            actual_content = f.read()
+        assert expected_content == actual_content
```
I wonder if it's worth it to check this bit-by-bit exactness? For a reader of the test, or the code, it's hard to see what the logic is - a test that would explain what's different when `use_first` is enabled (and test for that result) would probably be better...

(You could consider this comment repeated for all of the `assert expected_content == actual_content` snapshot tests.)

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        with open(self.output_file, 'r') as f:
+            actual_content = f.read()
+        assert expected_content == actual_content
```
Should we test that `read_po()` fails for this file? (Shouldn't it, given the new error there?)

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        fake_stdout, buf = self._capture_stdout()
+        with patch('sys.stdout', fake_stdout):
+            self.cmd.run()
```
[`contextlib.redirect_stdout`](https://docs.python.org/3.14/library/contextlib.html#contextlib.redirect_stdout) in general, but pytest has a [`capsys` fixture](https://docs.pytest.org/en/stable/how-to/capture-stdout-stderr.html) fir this exact purpose.

soft-suroleb replied:
capsys doesn't work here because the code writes to sys.stdout.buffer (binary stream), not sys.stdout — so monkeypatch is used instead.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        fake_stdout, buf = self._capture_stdout()
+        with patch('sys.stdout', fake_stdout):
+            self.cmd.run()
```
As above, re stdout capturing

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        self.temp_def = f'{i18n_dir}/msgmerge_def.po'
+        self.temp_ref = f'{i18n_dir}/msgmerge_ref.pot'
+        self.compendium = f'{i18n_dir}/compenidum.po'
```
Typo: compenidum

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+
+        self.dist = Distribution(TEST_PROJECT_DISTRIBUTION_DATA)
+        self.cmd = frontend.MergeCatalog(self.dist)
+        self.cmd.initialize_options()
```
This could become a `merge_cmd` fixture (see `extract_cmd`)

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            actual_content = f.read()
+        assert expected_content == actual_content
+
+    def test_compenidum(self):
```
Typo: compenidum

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            actual_content = f.read()
+        assert expected_content == actual_content
+
+    def test_compenidum_overwrite(self):
```
Typo: compenidum

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""msgid "word1"
```
```suggestion
        expected_content = self._get_expected("""msgid "word1"
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""#: simple.py:1
```
```suggestion
        expected_content = self._get_expected("""#: simple.py:1
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""#: simple.py:1
```
```suggestion
        expected_content = self._get_expected("""#: simple.py:1
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""#: simple.py:1
```
```suggestion
        expected_content = self._get_expected("""#: simple.py:1
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""#: hard.py:100 simple.py:100
```
```suggestion
        expected_content = self._get_expected("""#: hard.py:100 simple.py:100
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""msgid "word1"
```
```suggestion
        expected_content = self._get_expected("""msgid "word1"
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""msgid "word1"
```
```suggestion
        expected_content = self._get_expected("""msgid "word1"
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        expected_content = self._get_expected(fr"""msgid "word1"
```
```suggestion
        expected_content = self._get_expected("""msgid "word1"
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.finalize_options()
+        self.cmd.run()
+
+        assert os.path.exists(self.temp_def + '~')
```
```suggestion
```
Unnecessary assertion, the next line will fail with FileNotFoundError if it's not there.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+        self.cmd.suffix = '.bac'
+        self.cmd.run()
+
+        assert os.path.exists(self.temp_def + '.bac')
```
Ditto, unnecessary assertion.
```suggestion
```

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            assert 'msgstr "Word 3 from comp2"' in content
+        finally:
+            if os.path.exists(compendium2):
+                os.unlink(compendium2)
```
Try-finally would be unnecessary with tmp_path fixture.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            assert 'msgstr "Word 3 comp"' in content
+        finally:
+            if os.path.exists(compendium_with_word3):
+                os.unlink(compendium_with_word3)
```
Ditto, try-finally unnecessary with tmp_path.

akx commented on tests/messages/frontend/test_concat_merge.py:
```diff
@@ -0,0 +1,806 @@
+            assert '#~ msgstr "Old Word"' in content
+        finally:
+            if os.path.exists(extra_def):
+                os.unlink(extra_def)
```
Ditto, try-finally unnecessary with tmp_path.

akx reviewed (changes requested):
Thanks for addressing the previous comments!

Another round within - something I'd be extra careful about is those new APIs you're adding to `Catalog`; public APIs are hard to change neatly afterwards, so they should be designed with precision...

soft-suroleb committed:
Address review comments in catalog, frontend and pofile

- catalog: use _key_for(id, context) in add_conflict/get_conflicts to
  correctly handle messages with msgctxt; add docstrings to both methods
- frontend: tighten type hints (Counter[_MessageID], dict[_MessageID, ...]);
  remove redundant early-return guard in _get_messages_from_compendiums;
  initialize self.compendium as empty list
- pofile: replace .startswith() calls with slice comparisons for consistency

soft-suroleb committed:
Refactor tests: split test_concat_merge.py into test_concat.py and test_merge.py

- Use pytest fixtures (tmp_path, monkeypatch) instead of setup_method/teardown_method
- Use pathlib.Path for file paths
- Replace FakeStdout class with monkeypatch.setattr on sys.stdout
- Replace snapshot assertions with targeted behavioral checks
- Fix compendium typo (compenidum -> compendium)
- Fix self.compendium default to [] so _get_messages_from_compendiums is safe without guard
- Add test_conflicted_po_raises_on_read

soft-suroleb commented:
@akx @FrBrGeorge

Addressed all review comments from the latest round:
**catalog.py** — `add_conflict` now uses `_key_for(id, context)` so messages with `msgctxt` are keyed correctly; added docstrings to `add_conflict` and `get_conflicts`
**frontend.py** — tightened type hints (`Counter[_MessageID]`, `dict[_MessageID, ...]`); removed redundant early-return guard in `_get_messages_from_compendiums`; initialized `self.compendium` as `[]`
**pofile.py** — replaced `.startswith()` calls with slice comparisons per the codebase convention
**tests** — split `test_concat_merge.py` into `test_concat.py` and `test_merge.py`; converted to pytest style with `tmp_path`, `monkeypatch`, `pathlib.Path`; replaced snapshot assertions with targeted checks; fixed `compenidum` typo; added `test_conflicted_po_raises_on_read`
Left two comments for discussion:
- **pofile conflict handling** — `_invalid_pofile` is already a warning by default (only raises with `abort_invalid=True`), which is strictly better than the old behavior of silently parsing conflict markers as user comments
- **stdout capture** — `capsys` doesn't work here since the code writes to `sys.stdout.buffer` (binary), so `monkeypatch` is used instead
